cli/src/commands/parse/msgs.rs

```diff
@@ -0,0 +1,426 @@
+
+    #[inline]
+    fn add_uploaded(&self, num_uploaded: usize) {
+        self.uploaded.fetch_add(num_uploaded, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn increment_failed(&self) {
+        self.failed.fetch_add(1, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn increment_processed(&self) {
+        self.processed.fetch_add(1, Ordering::SeqCst);
+    }
+
+    #[inline]
+    fn num_uploaded(&self) -> usize {
+        self.uploaded.load(Ordering::SeqCst)
+    }
+
+    #[inline]
+    fn num_failed(&self) -> usize {
+        self.failed.load(Ordering::SeqCst)
+    }
+
+    #[inline]
+    fn num_processed(&self) -> usize {
+        self.processed.load(Ordering::SeqCst)
+    }
+}
+
+#[derive(Debug, StructOpt)]
+pub struct ParseMsgArgs {
+    #[structopt(short = "d", long = "dir", parse(from_os_str))]
+    /// Directory containing the msgs
+    directory: PathBuf,
+
+    #[structopt(short = "s", long = "source")]
+    /// Source name or id
+    source: SourceIdentifier,
+
+    #[structopt(long = "transform-tag")]
+    /// Transform tag to use.
+    transform_tag: TransformTag,
+
+    #[structopt(short = "n", long = "no-charge")]
+    /// Whether to attempt to bypass billing (internal only)
+    no_charge: bool,
+
+    #[structopt(short = "y", long = "yes")]
+    /// Consent to ai unit charge. Suppresses confirmation prompt.
+    yes: bool,
+}
+
+fn read_stream(stream_path: PathBuf, compound_file: &mut CompoundFile<File>) -> Result<Vec<u8>> {
+    let data = {
+        let mut stream = compound_file.open_stream(stream_path)?;
+        let mut buffer = Vec::new();
+        stream.read_to_end(&mut buffer)?;
+        buffer
+    };
+
+    Ok(data)
+}
+
+fn read_unicode_stream_to_string(
+    stream_path: PathBuf,
+    compound_file: &mut CompoundFile<File>,
+) -> Result<String> {
+    if !compound_file.is_stream(stream_path.clone()) {
+        return Err(anyhow!(
+            "Could not find stream {}. Please check that you are using unicode msgs",
+            stream_path.to_string_lossy()
+        ));
+    }
+
+    let data = read_stream(stream_path, compound_file)?;
+
+    // Stream data is a UTF16 string encoded as Vec[u8]
+    let mut buffer: String = String::with_capacity(data.len() * 2);
+    let mut decoder = encoding_rs::UTF_16LE.new_decoder();
+    let (coder_result, _, _) = decoder.decode_to_string(&data, &mut buffer, true);
+    if coder_result != encoding_rs::CoderResult::InputEmpty {
+        Err(anyhow!("Unexpected coder result"))
+    } else {
+        Ok(buffer)
+    }
+}
+
+fn get_attachment_store_path(attachment_number: usize) -> PathBuf {
+    PathBuf::from(format!(
+        "{}{:08}",
+        STREAM_PATH_ATTACHMENT_STORE_PREFIX, attachment_number
+    ))
+}
+
+fn read_attachment(
+    attachment_path: PathBuf,
+    compound_file: &mut CompoundFile<File>,
+) -> Result<AttachmentMetadata> {
+    let mut attachment_name_path = attachment_path.clone();
+    attachment_name_path.push(STREAM_PATH_ATTACHMENT_FILENAME.clone());
```

Comment:
nit: don't have to clone here
```suggestion
    attachment_name_path.push(&*STREAM_PATH_ATTACHMENT_FILENAME);
```

`*` dereferences your `Lazy` instance to the inner value, then `&` borrows the `PathBuf` into a `Path` (which is all you need for `push` here)